pkg/puller/puller.go

```diff
@@ -104,247 +100,154 @@ func (p *Puller) manage(warmupTime time.Duration) {
-	p.syncPeersMtx.Unlock()
-
-	syncCtx.Lock()
-	defer syncCtx.Unlock()
-
-	p.cursorsMtx.Lock()
-	c, ok := p.cursors[peer.ByteString()]
-	p.cursorsMtx.Unlock()
+func (p *Puller) syncPeer(ctx context.Context, peer *syncPeer, syncRadius uint8) error {
+	peer.Lock()
+	defer peer.Unlock()
 
-	if !ok {
-		cursors, err := p.syncer.GetCursors(ctx, peer)
+	if peer.cursors == nil {
+		cursors, err := p.syncer.GetCursors(ctx, peer.address)
 		if err != nil {
-			loggerV2.Debug("could not get cursors from peer", "peer_address", peer, "error", err)
-			p.syncPeersMtx.Lock()
-			delete(p.syncPeers[po], peer.ByteString())
-			p.syncPeersMtx.Unlock()
-
-			return
-			// remove from syncing peers list, trigger channel to find some other peer
-			// maybe blacklist for some time
+			return fmt.Errorf("could not get cursors from peer %s: %w", peer.address, err)
 		}
-		p.cursorsMtx.Lock()
-		p.cursors[peer.ByteString()] = peerCursors{created: time.Now(), cursors: cursors}
-		c = p.cursors[peer.ByteString()]
-		p.cursorsMtx.Unlock()
+		peer.cursors = cursors
 	}
 
-	// if length of returned cursors does not add up to
-	// what we expect it to be - dont do anything
-	if len(c.cursors) != int(p.bins) {
-		p.syncPeersMtx.Lock()
-		delete(p.syncPeers[po], peer.ByteString())
-		p.syncPeersMtx.Unlock()
-		return
+	if len(peer.cursors) != int(p.bins) {
+		return errCursorsLength
 	}
 
-	for bin, cur := range c.cursors {
-		if bin == 0 || uint8(bin) < d {
-			continue
+	for bin, cur := range peer.cursors {
+		if bin >= int(syncRadius) && !peer.isBinSyncing(uint8(bin)) {
+			p.syncPeerBin(ctx, peer, uint8(bin), cur)
 		}
-		p.syncPeerBin(ctx, syncCtx, peer, uint8(bin), cur)
 	}
+
+	return nil
 }
 
-func (p *Puller) syncPeerBin(ctx context.Context, syncCtx *syncPeer, peer swarm.Address, bin uint8, cur uint64) {
+// syncPeerBin will start historical and live syncing for the peer for a particular bin.
+// Must be called under syncPeer lock.
+func (p *Puller) syncPeerBin(ctx context.Context, peer *syncPeer, bin uint8, cur uint64) {
 	binCtx, cancel := context.WithCancel(ctx)
-	syncCtx.setBinCancel(cancel, bin)
+	peer.setBinCancel(cancel, bin)
 	if cur > 0 {
 		p.wg.Add(1)
-		go p.histSyncWorker(binCtx, peer, bin, cur)
+		go p.histSyncWorker(binCtx, peer.address, bin, cur)
 	}
 	// start live
 	p.wg.Add(1)
-	go p.liveSyncWorker(binCtx, peer, bin, cur)
+	go p.liveSyncWorker(binCtx, peer.address, bin, cur)
 }
 
 func (p *Puller) histSyncWorker(ctx context.Context, peer swarm.Address, bin uint8, cur uint64) {
-	loggerV2 := p.logger.V(2).Register()
+	loggerV2 := p.logger
+
+	defer p.wg.Done()
+	defer p.metrics.HistWorkerDoneCounter.Inc()
+
+	sleep := false
+
+	p.logger.Debug("histSyncWorker starting", "peer_address", peer, "bin", bin, "cursor", cur)
 
-	defer func() {
-		p.wg.Done()
-		p.metrics.HistWorkerDoneCounter.Inc()
-	}()
-	loggerV2.Debug("histSyncWorker starting", "peer_address", peer, "bin", bin, "cursor", cur)
 	for {
 		p.metrics.HistWorkerIterCounter.Inc()
+
+		if sleep {
```

Comment:
The value of `sleep` variable is never changed.

It appears that the retry functionality in histSyncWorker is not tested, according to the coverage report. It would be good to have this functionality tested as it seems fragile enough for accidental subtle bugs.